##### Next word prediction is a natural language processing (NLP) task that involves predicting the most likely word to follow a given sequence of words in a sentence or text. It's a subtask of language modeling, where the goal is to understand and generate human-like language patterns. This task is essential in various applications, including Virtual Keyboards,Text Completion,Grammar and Style Correction.
This project is a text generation task project to do this i am going to convert this data into lebbled data so i can applay the superwised machine learning techinques.

In [1]:
text="""Mohandas Karamchand Gandhi was born on October 2, 1869.
He was an Indian lawyer and political ethicist.Gandhi played a key role in India's independence from British rule.
Known as Mahātmā, meaning 'great-souled,' he employed nonviolent resistance and inspired civil rights movements globally.
Raised in Gujarat, India, Gandhi studied law in London. Later, he lived in South Africa for 21 years, where he first used nonviolent resistance for civil rights.
Returning to India in 1915, Gandhi became a leader in the Indian National Congress. He advocated for self-rule, easing poverty, women's rights, and religious amity.
He led campaigns, including the Dandi Salt March in 1930, challenging British policies. Despite imprisonment, he continued to champion India's independence.
However, his vision of a unified, religiously pluralistic India faced challenges in the 1940s with the demand for a separate Muslim state.
After India gained independence in 1947, communal violence erupted.
Gandhi, troubled by the strife, undertook hunger strikes for peace. He was assassinated on January 30, 1948, by Nathuram Godse, a Hindu nationalist.
Gandhi's birthday is celebrated as Gandhi Jayanti, and he is revered as the Father of the Nation in post-colonial India.
His life symbolizes nonviolence and political activism, leaving a lasting impact on the world.
"""

In [2]:
import pandas as pd
import numpy as np
import tensorflow as tf
from nltk.tokenize import sent_tokenize
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense
import warnings
warnings.filterwarnings('ignore')

In [3]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts([text])

In [4]:
tokenizer.word_index # each Unique word is converted into tokens

{'in': 1,
 'he': 2,
 'the': 3,
 'gandhi': 4,
 'a': 5,
 'and': 6,
 'india': 7,
 'for': 8,
 'was': 9,
 'on': 10,
 'independence': 11,
 'as': 12,
 'rights': 13,
 'indian': 14,
 'political': 15,
 "india's": 16,
 'british': 17,
 'rule': 18,
 'nonviolent': 19,
 'resistance': 20,
 'civil': 21,
 'to': 22,
 'his': 23,
 'of': 24,
 'by': 25,
 'is': 26,
 'mohandas': 27,
 'karamchand': 28,
 'born': 29,
 'october': 30,
 '2': 31,
 '1869': 32,
 'an': 33,
 'lawyer': 34,
 'ethicist': 35,
 'played': 36,
 'key': 37,
 'role': 38,
 'from': 39,
 'known': 40,
 'mahātmā': 41,
 'meaning': 42,
 "'great": 43,
 'souled': 44,
 "'": 45,
 'employed': 46,
 'inspired': 47,
 'movements': 48,
 'globally': 49,
 'raised': 50,
 'gujarat': 51,
 'studied': 52,
 'law': 53,
 'london': 54,
 'later': 55,
 'lived': 56,
 'south': 57,
 'africa': 58,
 '21': 59,
 'years': 60,
 'where': 61,
 'first': 62,
 'used': 63,
 'returning': 64,
 '1915': 65,
 'became': 66,
 'leader': 67,
 'national': 68,
 'congress': 69,
 'advocated': 70,
 'self'

In [5]:
len(tokenizer.word_index)   #Total unique words in vocablury

139

In [6]:
for  sentence in sent_tokenize(text):
    print(sentence)

Mohandas Karamchand Gandhi was born on October 2, 1869.
He was an Indian lawyer and political ethicist.Gandhi played a key role in India's independence from British rule.
Known as Mahātmā, meaning 'great-souled,' he employed nonviolent resistance and inspired civil rights movements globally.
Raised in Gujarat, India, Gandhi studied law in London.
Later, he lived in South Africa for 21 years, where he first used nonviolent resistance for civil rights.
Returning to India in 1915, Gandhi became a leader in the Indian National Congress.
He advocated for self-rule, easing poverty, women's rights, and religious amity.
He led campaigns, including the Dandi Salt March in 1930, challenging British policies.
Despite imprisonment, he continued to champion India's independence.
However, his vision of a unified, religiously pluralistic India faced challenges in the 1940s with the demand for a separate Muslim state.
After India gained independence in 1947, communal violence erupted.
Gandhi, troubled

In [7]:
for  sentence in text.split('\n'):
    print(tokenizer.texts_to_sequences([sentence])[0])

[27, 28, 4, 9, 29, 10, 30, 31, 32]
[2, 9, 33, 14, 34, 6, 15, 35, 4, 36, 5, 37, 38, 1, 16, 11, 39, 17, 18]
[40, 12, 41, 42, 43, 44, 45, 2, 46, 19, 20, 6, 47, 21, 13, 48, 49]
[50, 1, 51, 7, 4, 52, 53, 1, 54, 55, 2, 56, 1, 57, 58, 8, 59, 60, 61, 2, 62, 63, 19, 20, 8, 21, 13]
[64, 22, 7, 1, 65, 4, 66, 5, 67, 1, 3, 14, 68, 69, 2, 70, 8, 71, 18, 72, 73, 74, 13, 6, 75, 76]
[2, 77, 78, 79, 3, 80, 81, 82, 1, 83, 84, 17, 85, 86, 87, 2, 88, 22, 89, 16, 11]
[90, 23, 91, 24, 5, 92, 93, 94, 7, 95, 96, 1, 3, 97, 98, 3, 99, 8, 5, 100, 101, 102]
[103, 7, 104, 11, 1, 105, 106, 107, 108]
[4, 109, 25, 3, 110, 111, 112, 113, 8, 114, 2, 9, 115, 10, 116, 117, 118, 25, 119, 120, 5, 121, 122]
[123, 124, 26, 125, 12, 4, 126, 6, 2, 26, 127, 12, 3, 128, 24, 3, 129, 1, 130, 131, 7]
[23, 132, 133, 134, 6, 15, 135, 136, 5, 137, 138, 10, 3, 139]
[]


In [8]:
input_sequences = []
for sentence in sent_tokenize(text):
    tokenized_sentence = tokenizer.texts_to_sequences([sentence])[0]
    for i in range(1,len(tokenized_sentence)):
         input_sequences.append(tokenized_sentence[:i+1])

In [9]:
input_sequences

[[27, 28],
 [27, 28, 4],
 [27, 28, 4, 9],
 [27, 28, 4, 9, 29],
 [27, 28, 4, 9, 29, 10],
 [27, 28, 4, 9, 29, 10, 30],
 [27, 28, 4, 9, 29, 10, 30, 31],
 [27, 28, 4, 9, 29, 10, 30, 31, 32],
 [2, 9],
 [2, 9, 33],
 [2, 9, 33, 14],
 [2, 9, 33, 14, 34],
 [2, 9, 33, 14, 34, 6],
 [2, 9, 33, 14, 34, 6, 15],
 [2, 9, 33, 14, 34, 6, 15, 35],
 [2, 9, 33, 14, 34, 6, 15, 35, 4],
 [2, 9, 33, 14, 34, 6, 15, 35, 4, 36],
 [2, 9, 33, 14, 34, 6, 15, 35, 4, 36, 5],
 [2, 9, 33, 14, 34, 6, 15, 35, 4, 36, 5, 37],
 [2, 9, 33, 14, 34, 6, 15, 35, 4, 36, 5, 37, 38],
 [2, 9, 33, 14, 34, 6, 15, 35, 4, 36, 5, 37, 38, 1],
 [2, 9, 33, 14, 34, 6, 15, 35, 4, 36, 5, 37, 38, 1, 16],
 [2, 9, 33, 14, 34, 6, 15, 35, 4, 36, 5, 37, 38, 1, 16, 11],
 [2, 9, 33, 14, 34, 6, 15, 35, 4, 36, 5, 37, 38, 1, 16, 11, 39],
 [2, 9, 33, 14, 34, 6, 15, 35, 4, 36, 5, 37, 38, 1, 16, 11, 39, 17],
 [2, 9, 33, 14, 34, 6, 15, 35, 4, 36, 5, 37, 38, 1, 16, 11, 39, 17, 18],
 [40, 12],
 [40, 12, 41],
 [40, 12, 41, 42],
 [40, 12, 41, 42, 43],
 [40, 12, 4

In [10]:
max_len = max([len(x) for x in input_sequences])
max_len

22

In [11]:
from tensorflow.keras.preprocessing.sequence import pad_sequences
padded_input_sequences = pad_sequences(input_sequences, maxlen = max_len, padding='pre')

In [12]:
X = padded_input_sequences[:,:-1]
X.shape

(193, 21)

In [13]:
y = padded_input_sequences[:,-1]
y.shape

(193,)

### This is a classification problem so i need  to convert these 193 numbers into classes

In [14]:
#Doing one hot encoding of all the classes sinse we have 139 unique words i am making 140 classes cuz it strarts with zero class
y = to_categorical(y,num_classes=140)
y.shape

(193, 140)

In [15]:
model = Sequential()
model.add(Embedding(140, 150, input_length=21))
model.add(LSTM(150))
model.add(Dense(140, activation='softmax'))

In [16]:
model.compile(loss='categorical_crossentropy', optimizer='adam',metrics=['accuracy'])
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 21, 150)           21000     
                                                                 
 lstm (LSTM)                 (None, 150)               180600    
                                                                 
 dense (Dense)               (None, 140)               21140     
                                                                 
Total params: 222,740
Trainable params: 222,740
Non-trainable params: 0
_________________________________________________________________


In [17]:
model.fit(X,y,epochs=80)

Epoch 1/80
7/7 [==============================] - 4s 35ms/step - loss: 4.9407 - accuracy: 0.0155
Epoch 2/80
7/7 [==============================] - 0s 37ms/step - loss: 4.9086 - accuracy: 0.0725
Epoch 3/80
7/7 [==============================] - 0s 34ms/step - loss: 4.8259 - accuracy: 0.0518
Epoch 4/80
7/7 [==============================] - 0s 35ms/step - loss: 4.7284 - accuracy: 0.0259
Epoch 5/80
7/7 [==============================] - 0s 35ms/step - loss: 4.6910 - accuracy: 0.0518
Epoch 6/80
7/7 [==============================] - 0s 32ms/step - loss: 4.6553 - accuracy: 0.0518
Epoch 7/80
7/7 [==============================] - 0s 33ms/step - loss: 4.6296 - accuracy: 0.0518
Epoch 8/80
7/7 [==============================] - 0s 33ms/step - loss: 4.6245 - accuracy: 0.0622
Epoch 9/80
7/7 [==============================] - 0s 33ms/step - loss: 4.5986 - accuracy: 0.0518
Epoch 10/80
7/7 [==============================] - 0s 35ms/step - loss: 4.5621 - accuracy: 0.0466
Epoch 11/80
7/7 [============

In [18]:
def pred(text):


  for i in range(5):
    # tokenize
    token_text = tokenizer.texts_to_sequences([text])[0]
    # padding
    padded_token_text = pad_sequences([token_text], maxlen=21, padding='pre')
    # predict
    pos = np.argmax(model.predict(padded_token_text))

    for word,index in tokenizer.word_index.items():
      if index == pos:
        text = text + " " + word
        print(text)

In [19]:
pred('After')

1/1 [==============================] - 1s 732ms/step
After india
1/1 [==============================] - 0s 34ms/step
After india gained
1/1 [==============================] - 0s 51ms/step
After india gained independence
1/1 [==============================] - 0s 34ms/step
After india gained independence in
1/1 [==============================] - 0s 39ms/step
After india gained independence in 1947


In [20]:
pred('He was')

1/1 [==============================] - 0s 33ms/step
He was an
1/1 [==============================] - 0s 54ms/step
He was an indian
1/1 [==============================] - 0s 33ms/step
He was an indian lawyer
1/1 [==============================] - 0s 34ms/step
He was an indian lawyer and
1/1 [==============================] - 0s 34ms/step
He was an indian lawyer and political


In [21]:
pred('known')

1/1 [==============================] - 0s 33ms/step
known as
1/1 [==============================] - 0s 33ms/step
known as mahātmā
1/1 [==============================] - 0s 33ms/step
known as mahātmā meaning
1/1 [==============================] - 0s 33ms/step
known as mahātmā meaning 'great
1/1 [==============================] - 0s 33ms/step
known as mahātmā meaning 'great souled
